File: OrcScala/src/orc/oil/Transformer.scala
//
// Transformer.scala -- Transformer object for Oil.
// Project OrcScala
//
// $Id: Oil.scala 1652 2010-05-27 18:33:35Z sidney.rosario $
//
// Created by sidney.rosario on May 10, 2010.
//
// Copyright (c) 2010 The University of Texas at Austin. All rights reserved.
//
// Use and redistribution of this file is governed by the license terms in
// the LICENSE file found in the project's top-level directory and also found at
// URL: http://orc.csres.utexas.edu/license.shtml .
//
  
package orc.oil

object Transformer {

  import orc.oil._
  
  /* Tags the tail calls in the body of definitions
   * in the input expression. */
  def tagTailCalls(e: Expression) {
	
	  object TailCallMarker extends OILWalker {
	    private def markTails(e: Expression) {
	      e match {
	        case c@ Call(target, args, typeArgs) => c.isTailCall = true
	        case f || g => markTails(f); markTails(g) 
	        case f >> g => markTails(g)
	        case f << g => markTails(f)
	        case f ow g => markTails(f); markTails(g) // ?? Verify this.
	        case DeclareDefs(defs, body) => markTails(body) 
	        case HasType(body,typ) => markTails(body) 
	      }
		}
	     
		override def enter(d : Def) {
			markTails(d.body)
		}
	 }
	 
	 TailCallMarker walk e
  }
  
  /** A basic walker class for OIL expressions */
  class OILWalker {
	  def walk(e: Expression) {
		e match {
		  case ex@ Stop() => {
			  enter(ex)
			  leave(ex)
		  }
		  case ex@ Constant(a) => {
			  enter(ex)
			  leave(ex)
		  }
		  case ex@ Variable(i) => {
			  enter(ex)
			  leave(ex)
		  }
		  case ex@ Call(target, args, typeArgs) => {
			  enter(ex)
			  walk(target)
			  for(a <- args) walk(a)
			  leave(ex)
		  }
		  case ex@ Parallel(f, g) => {
			  enter(ex)
		 	  walk(f)
		 	  walk(g)
			  leave(ex)
		  }
		  case ex@ Sequence(f, g) => {
			  enter(ex)
		 	  walk(f)
		 	  walk(g)
			  leave(ex)
		  }
		  case ex@ Prune(f, g) => {
			  enter(ex)
		 	  walk(f)
		 	  walk(g)
			  leave(ex)
		  } 
		  case ex@ Otherwise(f, g) => {
			  enter(ex)
		 	  walk(f)
		 	  walk(g)
			  leave(ex)
		  } 
		  case ex@ DeclareDefs(defs, body) => {
			  enter(ex)
		 	  for(d <- defs) {
		 	 	  enter(d)
		 	 	  walk(d.body)
		 	 	  leave(d)
		 	  }
		 	  walk(body)
			  leave(ex)
		  } 
		  case ex@ HasType(body,typ) => {
			  enter(ex)
		 	  walk(body)
			  leave(ex)
		  } 
		}
	  }
	  
	  /* 
	   * Enter and Leave methods for specific expressions.
	   * The default behaviour does nothing. Walkers should
       * override these methods as desired.
       */
	  def enter(e: Stop) {}
	  def leave(e: Stop) {}
	  
	  def enter(e: Constant) {}
	  def leave(e: Constant) {}
	  
	  def enter(e: Variable) {}
	  def leave(e: Variable) {}
	  
	  def enter(e: Call) {}
	  def leave(e: Call) {}
	   
	  def enter(e: Parallel) {}
	  def leave(e: Parallel) {}
	  
	  def enter(e: Sequence) {}
	  def leave(e: Sequence) {}
	  
	  def enter(e: Prune) {}
	  def leave(e: Prune) {}
	   
	  def enter(e: Otherwise) {}
	  def leave(e: Otherwise) {}
	  
	  def enter(e: DeclareDefs) {}
	  def leave(e: DeclareDefs) {}
	   
	  def enter(e: HasType) {}
	  def leave(e: HasType) {}
	   
	  def enter(d: Def) {}
	  def leave(d: Def) {}
  	}
}

File: tags/v1.1.0/OrcJava/src/orc/ast/oil/visitor/Transformer.java
//
// Transformer.java -- Java class Transformer
// Project OrcJava
//
// $Id$
//
// Copyright (c) 2010 The University of Texas at Austin. All rights reserved.
//
// Use and redistribution of this file is governed by the license terms in
// the LICENSE file found in the project's top-level directory and also found at
// URL: http://orc.csres.utexas.edu/license.shtml .
//

package orc.ast.oil.visitor;

import java.util.LinkedList;

import orc.ast.oil.expression.Call;
import orc.ast.oil.expression.Catch;
import orc.ast.oil.expression.DeclareDefs;
import orc.ast.oil.expression.DeclareType;
import orc.ast.oil.expression.Def;
import orc.ast.oil.expression.Expression;
import orc.ast.oil.expression.HasType;
import orc.ast.oil.expression.Otherwise;
import orc.ast.oil.expression.Parallel;
import orc.ast.oil.expression.Pruning;
import orc.ast.oil.expression.Sequential;
import orc.ast.oil.expression.Stop;
import orc.ast.oil.expression.Throw;
import orc.ast.oil.expression.WithLocation;
import orc.ast.oil.expression.argument.Argument;
import orc.ast.oil.expression.argument.Constant;
import orc.ast.oil.expression.argument.Field;
import orc.ast.oil.expression.argument.Site;
import orc.ast.oil.expression.argument.Variable;
import orc.ast.oil.type.Type;

/**
 * Abstract base class tree transformer for Oil expressions.
 * 
 * @author quark
 */
public abstract class Transformer implements Visitor<Expression> {

	public Expression visit(final Parallel expr) {
		return new Parallel(expr.left.accept(this), expr.right.accept(this));
	}

	public Expression visit(final Call expr) {
		final LinkedList<Argument> newArgs = new LinkedList<Argument>();
		for (final Argument a : expr.args) {
			newArgs.add((Argument) a.accept(this));
		}
		LinkedList<Type> newTypeArgs = null;
		if (expr.typeArgs != null) {
			newTypeArgs = new LinkedList<Type>();
			for (final Type t : expr.typeArgs) {
				newTypeArgs.add(visit(t));
			}
		}
		return new Call((Argument) expr.callee.accept(this), newArgs, newTypeArgs);
	}

	public Type visit(final Type type) {
		return type;
	}

	public Expression visit(final Constant arg) {
		return arg;
	}

	public Expression visit(final DeclareDefs expr) {
		final LinkedList<Def> newDefs = new LinkedList<Def>();
		for (final Def d : expr.defs) {
			newDefs.add(visit(d));
		}
		return new DeclareDefs(newDefs, expr.body.accept(this));
	}

	public Def visit(final Def d) {
		LinkedList<Type> newArgTypes = null;
		if (d.argTypes != null) {
			newArgTypes = new LinkedList<Type>();
			for (final Type t : d.argTypes) {
				newArgTypes.add(visit(t));
			}
		}
		Type newResultType = null;
		if (d.resultType != null) {
			newResultType = visit(d.resultType);
		}
		return new Def(d.arity, d.body.accept(this), d.typeArity, newArgTypes, newResultType, d.location, d.name);
	}

	public Expression visit(final Field arg) {
		return arg;
	}

	public Expression visit(final HasType hasType) {
		return new HasType(hasType.body.accept(this), visit(hasType.type), hasType.checkable);
	}

	public Expression visit(final Pruning expr) {
		return new Pruning(expr.left.accept(this), expr.right.accept(this), expr.name);
	}

	public Expression visit(final Sequential expr) {
		return new Sequential(expr.left.accept(this), expr.right.accept(this), expr.name);
	}

	public Expression visit(final Otherwise expr) {
		return new Otherwise(expr.left.accept(this), expr.right.accept(this));
	}

	public Expression visit(final Stop expr) {
		return expr;
	}

	public Expression visit(final Site arg) {
		return arg;
	}

	public Expression visit(final DeclareType typeDecl) {
		return new DeclareType(visit(typeDecl.type), typeDecl.body.accept(this));
	}

	public Expression visit(final Variable arg) {
		return arg;
	}

	public Expression visit(final WithLocation expr) {
		return new WithLocation(expr.body.accept(this), expr.location);
	}

	public Expression visit(final Throw throwExpr) {
		return new Throw(throwExpr.exception.accept(this));
	}

	public Expression visit(final Catch catchExpr) {
		return new Catch(this.visit(catchExpr.handler), catchExpr.tryBlock.accept(this));
	}
}
